`Single::compressAssociative()` method has been improved

Smoren opened the pull request:
Hi @markrogoyski,

I've improved `Single::compressAssociative()` method with more quick separate logic for arrays and `ArrayAccess` instances.

Now this method gets values from iterable by keys using `[]` operator when it can.

Smoren committed:
`Single::compressAssociative()` method has been improved with more quick separate logic for arrays and `ArrayAccess` instances.

Smoren commented:
This branch is rebase from `develop`.

Smoren commented:
I've rebase this branch from `develop` again.

markrogoyski commented on tests/Fixture/ArrayAccessIteratorFixture.php:
```diff
@@ -0,0 +1,60 @@
+
+namespace IterTools\Tests\Fixture;
+
+class ArrayAccessIteratorFixture implements \ArrayAccess, \Iterator
```
Question: Is this class necessary? Could the tests just use the built-in `\ArrayIterator` as that also implements `\ArrayAccess` and `\Iterator`? Or is there something else this fixture iterator provides that the built-in iterator does not?

Smoren replied:
Done: caa0b57d48d9484fdc1fc77c5a497a164ae45d9a.

Smoren committed:
`ArrayAccessIteratorFixture` removed and replaced by `ArrayIterator`.

Smoren commented:
Hi @markrogoyski 

Thank you for this note. I didn't know that `Arrayiterator` implements `ArrayAccess`. I removed the extra fixture and replaced its use with `ArrayIterator`.

markrogoyski commented on tests/Single/CompressAssociativeTest.php:
```diff
@@ -125,14 +125,14 @@ public function dataProviderForArray(): array
                 ['a', 'b', 0, 2],
-                [0, 2, 'a', 'b'],
-                [1, 3, 11, 22],
+                ['a', 'b', 0, 2],
+                [11, 22, 1, 3],
```
I think this is an unexpected and undesirable result which changes the definition of the action and what an iterable is. In [PHP, array's](https://www.php.net/manual/en/language.types.array.php) are not dictionaries/maps like which are found in other languages. It is technically an ordered map (or ordered dictionary). The compress function is expected to filter out specific indexes without changing the order of the elements. The expectation would be the same result, regardless if the input was an array, iterator, or generator, etc.

For reference, Python has a similar data structure: [OrderedDict](https://docs.python.org/3/library/collections.html#collections.OrderedDict).

Smoren replied:
I think performance is more important here than order. The main thing is to get exactly the right elements for the right keys.

markrogoyski replied:
Hi @Smoren,

Thank you for considering how to improve performance. However, performance doesn't matter if it gets the wrong answer. PHP arrays are an ordered map, so the order is fundamentally part of the data structure. If the order changes, the result does not match the input.

Let me demonstrate with two simple examples:

This first example shows intent, and compress subverting expectations:
```php
$soonToBeSorted = ['a' => 4, 'b' => 5, 'c' => 6, 1, 2, 3];  // [4, 5, 6, 1, 2, 3]
asort($soonToBeSorted);                                     // [1, 2, 3, 4, 5, 6]

foreach (Single::compressAssociative($soonToBeSorted, ['a', 'b', 0, 2]) as $x) {
    print($x . ", ");  // 4, 5, 1, 3
}
```
In this example, clearly a sorted result is expected, but compress changed the order to no longer be sorted.

This second example shows a real scenario of data where order matters:
```php
$chessMoves = [
    'whiteOpening' => 'f3',
    'blackOpening' => 'e6', 
    'whiteMidGame' => 'g4',
    'checkmate'    => 'Qh4'
];

$endGameCompressor = ['checkmate', 'whiteMidGame'];

foreach (Single::compressAssociative($chessMoves, $endGameCompressor) as $move => $notation) {
    print("$move: $notation\n");
}

// checkmate: Qh4
// whiteMidGame: g4
```

In this example, we just want to display the interesting part of the game, but compress changes the order, so you can't make sense of the chess game anymore (a real non-fool's mate would have a longer end game with more moves).

Order is fundamental to PHP arrays as well as iterators and generators. Changing the order would be unexpected and most likely result in user complaints, even if it was documented.

Smoren replied:
OK, you are true. I'm closing this PR.
